src/snowflake/snowpark/dataframe_reader.py

```diff
@@ -1024,3 +1073,480 @@ def _read_semi_structured_file(self, path: str, format: str) -> DataFrame:
+            )
+            self._session.sql(sql_create_temp_stage).collect(
+                statement_params=statements_params_for_telemetry
+            )
+
+            try:
+                with mp.Manager() as process_manager, ProcessPoolExecutor(
+                    max_workers=max_workers
+                ) as process_executor, ThreadPoolExecutor(
+                    max_workers=max_workers
+                ) as thread_executor:
+                    thread_pool_futures, process_pool_futures = [], []
+                    parquet_file_queue = process_manager.Queue()
+
+                    def ingestion_thread_cleanup_callback(parquet_file_path, _):
+                        # clean the local temp file after ingestion to avoid consuming too much temp disk space
+                        shutil.rmtree(parquet_file_path, ignore_errors=True)
+
+                    logger.info("Starting to fetch data from the data source.")
+                    for partition_idx, query in enumerate(partitioned_queries):
+                        process_future = process_executor.submit(
+                            DataFrameReader._task_fetch_from_data_source_with_retry,
+                            parquet_file_queue,
+                            create_connection,
+                            query,
+                            struct_schema,
+                            partition_idx,
+                            tmp_dir,
+                            dbms_type,
+                            query_timeout,
+                            fetch_size,
+                            session_init_statement,
+                        )
+                        process_pool_futures.append(process_future)
+                    # Monitor queue while tasks are running
+                    while True:
+                        try:
+                            file = parquet_file_queue.get_nowait()
+                            logger.debug(f"Retrieved file from parquet queue: {file}")
+                            thread_future = thread_executor.submit(
+                                self._upload_and_copy_into_table_with_retry,
+                                file,
+                                snowflake_stage_name,
+                                snowflake_table_name,
+                                "abort_statement",
+                                statements_params_for_telemetry,
+                            )
+                            thread_future.add_done_callback(
+                                functools.partial(
+                                    ingestion_thread_cleanup_callback, file
+                                )
+                            )
+                            thread_pool_futures.append(thread_future)
+                            logger.debug(
+                                f"Submitted file {file} to thread executor for ingestion."
+                            )
+                        except queue.Empty:
```

Comment:
Do you create all parquet files, put them in the queue, then upload and process them?

Reply:
I believe so, @sfc-gh-aling can confirm

Reply:
it's parallel, we use producer-consumer model -- once a file is created, the path will be added to a queue, the ingestion consumer will get the file from the queue to ingestion.